### ISSUE TITLE
Show actual memory usage for arrangements

### ISSUE THREAD

**Issue by rjnn** [MaterializeInc/materialize#1532]
Currently, the logging view `mz_arrangement_sizes` shows me a record count:

```
arjun=> select * from mz_arrangement_sizes;
 operator | worker | records | batches 
----------+--------+---------+---------
     1460 |      0 |  636249 |      15
     1475 |      0 |   50629 |      10
     1499 |      0 |  610501 |       9
     1503 |      0 |   50629 |       8
     1514 |      0 |      61 |       3
(5 rows)
```

While this is nice, I would also like to see the actual sizes of those arrangements in bytes as well. A record that contains a fat JSON blob is more problematic than a lot of ints, and currently I don't quite know what's in what.

**Comment by danhhz** [comment]
@frankmcsherry Do you have any thoughts about how to implement this?

I tried to pick this off as a skunkworks friday project, but none of the approaches I tried were satisfying. First, I'm assuming we want to keep this info maintained, instead of computing it in response to a request. It feels like having to walk an entire arrangement each time we want this isn't going to scale. The intuition I was hoping to use is that while all the Batches in a Trace are abstract over K,V,T,R, and it's unclear what the "size" of K, V, etc is, in materialize we've got them concretely specified as things that have well known sizes.

My first attempt was to create wrappers impls in mz for Batch, Builder, Merger, etc, that were specialized for our types and internally added a `byte_size: usize` that was computed as a batch was built/merged, but otherwise proxied all the logic. It's been a couple weeks, but IIRC this was a lot of boilerplate (even more than I'd expected) and kinda clunky along the mz/dd boundaries in a way that I might have been able to clean up with some elbow grease.

My second attempt was to add an additional trait constraint in dd for K, V, T, R that they be able to specify their own size in bytes. I ended up having to put this constraint in even more places than I expected and paused because I wasn't sure if the complexity was palatable.

Maybe there's some better third way I'm not seeing?

**Comment by frankmcsherry** [comment]
> @frankmcsherry Do you have any thoughts about how to implement this?

Mostly that it probably isn't as important as folks think. Our perf debugging so far has been based on record counts, and those seem to be the ones that move the needle the most (that is: maintaining too many records has been a better indicator of problems than "I am surprised by what is in my data"). There is probably an observability bias here.

https://github.com/MaterializeInc/materialize/pull/6999 was added recently, which is probably a fine way to define a MV to track expectations about row sizes, though without tracking various metadata overheads. 

https://github.com/MaterializeInc/materialize/pull/6763 is probably a fine way to get at arrangement sizes in bytes, as well as a breakdown of other associated metadata. Until we do something like this, I'm skeptical of numbers that come back without e.g. memory fragmentation, distribution over NUMA regions, stuff like that.

**Comment by teskje** [comment]
@antiguru can this be closed now that #18766 has been merged?
